l0_sampling/sketch.cpp

```diff
@@ -1,52 +1,68 @@
-#include <math.h>
 #include <cmath>
-#include <assert.h>
+#include <cassert>
 #include <vector>
 #include <iostream>
 #include <string>
 #include "../include/prime_generator.h"
 #include "../include/sketch.h"
-#include "../include/util.h"
 
-Sketch::Sketch(uint64_t n, long seed): seed(seed), n(n), random_prime
-(PrimeGenerator::generate_prime(n*n)) {
-  const unsigned long long int num_buckets = double_to_ull(log2(n)+1);
-  buckets = std::vector<Bucket>(num_buckets*(log2(n)+1));
-  for (unsigned i = 0; i < num_buckets; ++i){
-    for (unsigned j = 0; j < log2(n)+1; ++j){
-      buckets[i*(log2(n)+1)+j].set_guess(1 << j);
-      buckets[i*(log2(n)+1)+j].set_seed(i*(log2(n)+1)+j,seed,random_prime);
-    }
-  }
+Sketch::Sketch(uint64_t n, long seed): seed(seed), n(n), large_prime
+(PrimeGenerator::generate_prime((uint128_t)n*n)) {
+  const unsigned long long int num_buckets = bucket_gen(n);
+  const unsigned long long int num_guesses = guess_gen(n);
+  buckets = std::vector<Bucket_Boruvka>(num_buckets * num_guesses);
 }
 
 void Sketch::update(Update update ) {
-  for (unsigned j = 0; j < buckets.size(); j++){
-    if (buckets[j].contains(update.index+1)){
-      buckets[j].a += update.delta;
-      buckets[j].b += update.delta*(update.index+1); // deals with updates whose indices are 0
-      buckets[j].c += (update.delta*PrimeGenerator::power(buckets[j].r,update.index+1,random_prime))%random_prime;
-      buckets[j].c = (buckets[j].c + random_prime)%random_prime;
+  const unsigned long long int num_buckets = bucket_gen(n);
+  const unsigned long long int num_guesses = guess_gen(n);
+  for (unsigned i = 0; i < num_buckets; ++i) {
+    for (unsigned j = 0; j < num_guesses; ++j) {
+      unsigned bucket_id = i * num_guesses + j;
+      XXH64_hash_t bucket_seed = XXH64(&bucket_id, sizeof(bucket_id), seed);
+      int128_t r = 2 +  bucket_seed % (large_prime - 3);
+      if (buckets[bucket_id].contains(update.index+1, bucket_seed, 1 << j)){
+        buckets[bucket_id].a += update.delta;
+        buckets[bucket_id].b += update.delta*(update.index+1); // deals with updates whose indices are 0
+        buckets[bucket_id].c = static_cast<uint128_t>(
+              (static_cast<int128_t>(buckets[bucket_id].c)
+              + static_cast<int128_t>(large_prime)
+              + (update.delta*PrimeGenerator::power(r,(uint128_t) update
+              .index+1, large_prime) % static_cast<int128_t>(large_prime)))
```

Comment:
Just to make sure I'm not going crazy, what is the +1 for here?

Reply:
You are correct, the +1 is to deal with the 0 index issue we had before.